### PR TITLE
Revamp Unit Tooltips

### PR DESCRIPTION
**Pull request by ron-murhammer**
## Overview
Addresses https://forums.triplea-game.org/topic/942/unit-tooltip-improvements-poll

## Functional Changes
Makes unit tooltips and unit list have a more structured and easier to read format.

## Manual Testing Performed
Tested on a number of maps and save games to see the changes.

## Before & After Screen Shots
Many examples here: https://forums.triplea-game.org/topic/942/unit-tooltip-improvements-poll/44

### Before
image

### After
image

### CODE REVIEW

**Comment by DanVanAtta** [comment]
@ron-murhammer I like the update, but am not sure the pipe delimited format makes that line easier to read. The `2x6` seems really hard to understand now compared to before. Have you considered keeping the labels & values paired together, like: ```Attack: 2, Defense: 2; Movement: 2```

I suspect with shortened and standardized labels, we could get the same simplicity. Going to a civ style `attack-defense-move` I do no think makes sense.

**Comment by DanVanAtta** [comment]
Hmm, I"m afraid I'm less sold, this does not make sense to me:

screenshot from 2018-07-22 18-10-17

Are these defensive powers, attack power? Where and what is that slash representing?

**Comment by ron-murhammer** [comment]
@DanVanAtta Well, we already use the attack/defense/move format in the purchase screen. There are different opinions on how verbose vs short hand. Those 3 are the core properties that every TripleA map uses. I'd encourage you to read the thread I linked if you want some background and different views.

Its both otherwise it would say "Support-Attack" or "Support-Defense". Open to a different format if you have some suggestions. Could replace the "/" with a "&".

**Comment by DanVanAtta** [comment]
I did check out the forum post. I'm sorry I missed it until recently. On the whole I think it's really good but there are a few things I would change.

On the purchase screen it looks okay because there are no labels, eg:
screenshot from 2018-07-22 20-07-42

With the labels, my eyes are jumping back and forth. You could maybe solve it by going to a column format, eg:
```
Attack |  Defense | Movement
  2      |    2      |     2
```

But keeping the label/values side-by-side seems simplest to me... 


**************

Thanks for clarifying:  "Support-Attack" or "Support-Defense"

I think it is most clearly stated as: 'attacker-support-bonus'. That might be too wordy though. We do have some space on this screen, in general I would favor for us to prefer clarity here over compactness .

File: game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java [hunk 1]
```diff
@@ -2734,410 +2735,399 @@ public String allUnitStatsForExporter() {
...
+        result.append(": <b>").append(tuple.getSecond()).append("</b>");
+      }
+      result.append("<br />");
+    }
+    return result.toString();
+  }
+
+  private static <T extends DefaultNamed> void addIntegerMapDescription(final String key,
+      final IntegerMap<T> integerMap, final List<Tuple<String, String>> tuples) {
+    if (integerMap != null && integerMap.size() > 0) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: I really like that you extracted these blocks, but I would nitpick that the control flow, the `if` statement should have stayed with the client code block. With this extraction, when reading the main block, you're going to have to jump down 300 lines to find out this helper method maybe was a no-op. Also to construct the entire control flow you need to consider all the helper methods.

**Comment by DanVanAtta** [comment]
To summarize: after reviewing this over, I've just the one nit over cleanup; otherwise looks solid.  The nit is not a blocker, but I do think it would help and would make this update even better.

File: game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java (hunk 1 above)
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: @DanVanAtta Funny enough that is what I did first and then didn't really like the repetitive if blocks 4 times in a row in the client block so moved it into the helper method as well. I think you could also argue to make the helper methods return the tuple rather than adding to the list of tuples to have a cleaner input/output for the method. If you did that though then you'd have another if check on if the result is present before adding it or add logic to skip empty tuples later on. I think both of those points are kind of a wash in this situation so would rather just move forward with this.

Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: Interesting points. I would tend to agree. 

(side note: FWIW, the cleaner input/ouput is a good step, but seems like that is the place to stop without finding a better solution)

Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java]: Sorry, that may not have been very clear.

- The refactoring logic you walk through sounds reasonable, it is an interesting conclusion.
- It seems a number of problems come up when we have the potentially empty return type. My 2 cents, keep the extracted functions as void, find a better way to solve further problems if the code could still use it (probably could).

In general though tend to agree we are at diminishing returns, it is pretty interesting your first approach took that path.
